### ISSUE TITLE
Refactor hyperoptimize

### ISSUE THREAD

**Issue by erikr** [aguirre-lab/ml4c3#1133]
## What and why

1. Currently, the `hyperoptimize` recipe randomly selects a hyperparameter combination from all possible combinations. If that combination was already performed, it skips doing it again, but still counts as an evaluation (which is limited by `--max_evals`). Ideally, if there are `c` unique hyperparameter combinations, this recipe should iterate over each combination exactly once.

    If `c > max_evals`, then the recipe should iterate over up to `max_evals` combinations, randomly selected from all `c` combinations.

    Here is an example `metrics_and_hyperparameters.csv` generated with `--max_evals 10` but only two possible unique hyperparameter combinations:
    ```
    | Trial | auc-train-sts_death | auc-test-sts_death | test_loss | train_loss | valid_loss | parameter_count | learning_rate |
    |-------|---------------------|--------------------|-----------|------------|------------|-----------------|---------------|
    | 0     | 0.605               | 0.686              | 3.38      | 4.84       | 3.41       | 2078            | 0.001         |
    | 1     | nan                 | nan                | 9.00E+09  | 9.00E+09   | 9.00E+09   | 0               | 0.001         |
    | 2     | nan                 | nan                | 9.00E+09  | 9.00E+09   | 9.00E+09   | 0               | 0.001         |
    | 3     | nan                 | nan                | 9.00E+09  | 9.00E+09   | 9.00E+09   | 0               | 0.001         |
    | 4     | nan                 | nan                | 9.00E+09  | 9.00E+09   | 9.00E+09   | 0               | 0.001         |
    | 5     | 0.453               | 0.489              | 5.21      | 6.9        | 5.31       | 2078            | 0.0001        |
    | 6     | nan                 | nan                | 9.00E+09  | 9.00E+09   | 9.00E+09   | 0               | 0.0001        |
    | 7     | nan                 | nan                | 9.00E+09  | 9.00E+09   | 9.00E+09   | 0               | 0.001         |
    | 8     | nan                 | nan                | 9.00E+09  | 9.00E+09   | 9.00E+09   | 0               | 0.0001        |
    | 9     | nan                 | nan                | 9.00E+09  | 9.00E+09   | 9.00E+09   | 0               | 0.0001        |
    ```

    If a duplicate hyperparameter configuration is encountered, it would be ideal to not add anything to the dictionary of model results. So the above CSV should not have any `nans`.

1. Allocate models across GPUs 1:1.
    Eventually we want multi-GPU support, e.g. if you have four GPUs available and allocated to hyperoptimize, and 10 models, iterate over 10 models, and use all available GPUs to train each model.

1. Accepts a `configuration.json` file that specifies the hyperparameter space to optimize over.

1. Works for both supervised learning (`train`) modes, RL, and clustering.

## Solution(s)
We can rewrite a hyperoptimization script to:
1. load a `hyperparameters.json` file
2. iterate over all unique combinations of hyperparameters in random order (random search is good enough; not sure fancier search algorithms are worth the time)
3. for each combo, overwrite values inherited from the baseline parameters used in the `hyperparameters.sh` script

We could also potentially use elements from [`dispatch.py`](https://github.com/aguirre-lab/ml4c3/blob/master/scripts/dispatch.py) to accomplish multi-GPU support (https://github.com/aguirre-lab/ml4c3/issues/1132).

**Comment by erikr** [comment]
@raimonpv @StevenSong thoughts?

**Comment by StevenSong** [comment]
even though we're using `hyperopt` package, aren't we currently only relying on random combination right now? if so, why do we even need the `hyperopt` package? can't we just generate a unique list of hyperparameter combinations using numpy/list operations, shuffle it, and try `--max_evals` combos?

**Comment by erikr** [comment]
Currently we are using "tree of parzen estimators", not random search:
https://github.com/aguirre-lab/ml4c3/blob/f679d4b06280d6d2f266fc7a6506ab5645a0583d/ml4c3/hyperoptimizers.py#L268

http://hyperopt.github.io/hyperopt/#algorithms

I don't think TPE is much better than random search though. Hence my suggestion above to not use `hyperopt` which we seem to agree on.

@raimonpv ?

**Comment by erikr** [comment]
@RogerPallares @raimonpv @StevenSong Do we want to brainstorm over a Zoom call, or share ideas here asynchronously? I'm glad to do whatever you all think is best.

**Comment by RogerPallares** [comment]
I'm ok with both, maybe a Zoom call can be more efficient? At least to share first ideas and be on the same page, and then we can iterate asynchronously here?
